### ISSUE TITLE
Fedora 40 --with-ublk build succeeds ublk methods missing

### ISSUE THREAD

**Issue by emphaticism** [spdk/spdk#3499]
# Sighting report

when passing --with-ublk at configuration build of spdk succeeds but ublk methods are not present in completed build

## Expected Behavior

`sudo ./scripts/rpc.py ublk_create_target`
should return with success

## Current Behavior

```
sudo ./scripts/rpc.py ublk_create_target
request:
{
  "method": "ublk_create_target",
  "req_id": 1
}
Got JSON-RPC error response
response:
{
  "code": -32601,
  "message": "Method not found"
}
```

## Possible Solution

Unknown, attempted removing the distro default installed liburing 2.5 similar to what was done https://github.com/spdk/spdk/issues/3282 but the issue wasn't resolved

Unsure if it's ideal to be pinning the default pkgdep to liburing 2.2 when recent distro releases are shipping 2.5+ I'll open a separate issue related to pkgdep assumptions that impact rhel & debian based distros

Sorry if I'm making any assumptions here that are off base I've done what research I can but may have overlooked past work or issues related to this topic

## Steps to Reproduce

```
sudo yum -y update
sudo yum -y install git
reboot

git clone https://github.com/spdk/spdk --recursive
cd spdk
sudo -E scripts/pkgdep.sh --all
sudo modprobe ublk_drv
sudo ./configure --with-ublk
sudo make -j$(nproc)
./test/unit/unittest.sh
#modify ulimits to allow for needed memlock size
sudo scripts/setup.sh
sudo ./build/bin/nvmf_tgt &
sudo ./scripts/rpc.py bdev_nvme_attach_controller -b bdev0 -t PCIe -a 0000:01:00.0
sudo ./scripts/rpc.py bdev_nvme_attach_controller -b bdev1 -t PCIe -a 0000:0b:00.0
sudo ./scripts/rpc.py ublk_create_target
```

build logs show ublk is being built as far as I can tell

```
grep -i ublk ../build.log
  CC lib/ublk/ublk.o
  CC lib/ublk/ublk_rpc.o
  LIB libspdk_ublk.a
  CC module/event/subsystems/ublk/ublk.o
  LIB libspdk_event_ublk.a
  TEST_HEADER include/spdk/ublk.h
  CXX test/cpp_headers/ublk.o
```

I also attempted this build on ubunut 24.04 using basically the same build process to the same effect, I'll be opening another issue for that as there were other issues with that build that might need to be addressed independent of this issue.

## Context (Environment including OS version, SPDK version, etc.)

```
uname -a
Linux build-box 6.10.5-200.fc40.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Aug 14 15:49:44 UTC 2024 x86_64 GNU/Linux
```
```
grep CONFIG_BLK_DEV_UBLK /boot/config-6.10.5-200.fc40.x86_64
CONFIG_BLK_DEV_UBLK=m
```
```
yum list installed liburing
Installed Packages
liburing.x86_64   2.5-3.fc40   @fedora
```

spdk version main branch commit [27134348c](https://github.com/spdk/spdk/commit/27134348c3842e65e44e58a670d02a4e5d8d774b)

I've also attached logs of running `sudo -E scripts/pkgdep.sh --all` `./configure --with-ublk` `sudo make -j$(nproc)` & `./test/unit/unittest.sh`

[pkgdep-install.log](https://github.com/user-attachments/files/16713740/pkgdep-install.log)
[configure.log](https://github.com/user-attachments/files/16713739/configure.log)
[unittest.log](https://github.com/user-attachments/files/16713742/unittest.log)
[build.log](https://github.com/user-attachments/files/16713738/build.log)

**Comment by mikeBashStuff** [comment]
> Unsure if it's ideal to be pinning the default pkgdep to liburing 2.2 when recent distro releases are shipping 2.5+

This pin was put in place as a request for making sure we use releases that are tested and stable (see https://github.com/spdk/spdk/issues/1673). If you prefer to use any other version then there's no need to depend on the pkgdep.

That said, we already have f40 build in our staging and it's working just fine in regards to `ublk` (checked both liburing versions, 2.2 and 2.5).  The problem may be with the application that you are actually testing - if you check lib dependencies for `nvmf_tgt` you can see `ublk` is simply not there. It is included in `spdk_tgt`'s list though, so the RPCs should work just fine with that particular app.  You can also try:

```
diff --git a/app/nvmf_tgt/Makefile b/app/nvmf_tgt/Makefile
index c0f48e773..bba74b92d 100644
--- a/app/nvmf_tgt/Makefile
+++ b/app/nvmf_tgt/Makefile
@@ -18,7 +18,7 @@ SPDK_LIB_LIST += env_dpdk_rpc
 endif

 ifeq ($(OS),Linux)
-SPDK_LIB_LIST += event_nbd
+SPDK_LIB_LIST += event_nbd event_ublk
 endif
```

**Comment by mikeBashStuff** [comment]
Also, one sidenote here:

The `install_liburing()` from our pkgdep quite naively assumes there's no other `liburing` version around. Essentially, when `liburing`'s `install` target puts proper symlinks in place, the plain `ldconfig` call will happily overwrite them in case there's still a newer version available. So this `/usr/lib64/liburing.so.2 -> /usr/lib64/liburing.so.2.2` may become again `/usr/lib64/liburing.so.2 ->  /usr/lib64/liburing.so.2.something_else`. Slapping `-X` could help, but if any entity dares to run plain `ldconfig` later on, for whatever reason, these symlinks may still change.

**Comment by emphaticism** [comment]
I just broke out the pkgdeps script and installed exactly what was needed, that along with using the correct app target `spdk_tgt` has allowed things to build and work as expected on both fedora and ubuntu.
Ubuntu required building liburing to set the libdir path as spdk expects it.

```
[2024-08-22 16:55:06.378026] ublk.c: 572:ublk_ctrl_cmd_get_features: *NOTICE*: User Copy enabled
[2024-08-22 16:55:06.378049] ublk.c: 758:ublk_create_target: *NOTICE*: UBLK target created successfully
```

Sorry for the erroneous report, I had presumed pkgdeps was a bit more intelligent about defaulting to using distro provided tooling when possible and falling back to building libraries if suitable installs were not provided by the distro. Defaulting to as reproducible a setup as possible seems very reasonable to me, I may have missed this being called out in the docs.
